package/android/src/main/java/com/mrousavy/camera/core/CameraSession.kt

```diff
@@ -479,12 +479,20 @@ class CameraSession(
     )
 
     // Zoom
+    val cameraCharacteristics = cameraManager.getCameraCharacteristics(cameraId!!)
+    val zoomRange = (
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+        cameraCharacteristics.get(CameraCharacteristics.CONTROL_ZOOM_RATIO_RANGE)
+      } else {
+        null
+      }
+    ) ?: Range(1f, cameraCharacteristics.get(CameraCharacteristics.SCALER_AVAILABLE_MAX_DIGITAL_ZOOM) ?: 1f)
+    val zoomClamped = zoomRange.clamp(zoom)
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-      captureRequest.set(CaptureRequest.CONTROL_ZOOM_RATIO, zoom)
+      captureRequest.set(CaptureRequest.CONTROL_ZOOM_RATIO, zoomClamped)
     } else {
-      val cameraCharacteristics = cameraManager.getCameraCharacteristics(cameraId!!)
       val size = cameraCharacteristics.get(CameraCharacteristics.SENSOR_INFO_ACTIVE_ARRAY_SIZE)!!
-      captureRequest.set(CaptureRequest.SCALER_CROP_REGION, size.zoomed(zoom))
+      captureRequest.set(CaptureRequest.SCALER_CROP_REGION, size.zoomed(zoomClamped))
     }
```

Comment:
This entire thing is the same code as in `CameraDevice+createPhotoCaptureRequest` - let's move that to one place to avoid duplication.

Please create a file in `extensions/` called `CaptureRequest+setZoom.kt`, and add a Kotlin extension function like so:

```kt
fun CaptureRequest.setZoom(zoom: Double, cameraCharacteristics: CameraDeviceCharacteristics) {
  ...
```

Then implement the thing there.